### PR TITLE
Add support for zooming with a focus

### PR DESCRIPTION
**Pull request by levinli303**
Add an option ENABLE_FOCUS_ZOOMING to support zooming (shift + left drag) with a focus by changing orientation during zooming so that the focus location (where the drag started) is kept where it was.

Also adding support for mobile pinch gesture based on this. Pinch on mobile is basically a focus and scale so it can be easily translated in to zooming with focus.

### CODE REVIEW

**Comment by 375gnu** [comment]
Could you provide a test build for mobile users?

**Comment by levinli303** [comment]
> Could you provide a test build for mobile users?

I intend to provide beta testing via Google Play/App Store after this change is merged.

File: src/celestia/celestiacore.h [hunk 1]
```diff
@@ -401,11 +402,18 @@ class CelestiaCore // : public Watchable<CelestiaCore>
     void setLayoutDirection(celestia::LayoutDirection);
 
 private:
+    struct MouseLocation
```
Comment (375gnu) [inline src/celestia/celestiacore.h:405]: Actually, this type is not required, please use `Eigen::Vector2f` instead.

File: src/celestia/celestiacore.cpp [hunk 2]
```diff
@@ -2267,15 +2254,59 @@ Eigen::Vector3f CelestiaCore::getPickRay(float x, float y, const celestia::View
...
+    observer->setFOV(newFOV);
+    setZoomFromFOV();
+
+    if (oldPickRay.has_value())
+    {
+        Eigen::Vector3f newPickRay = getPickRay(focus.value().x, focus.value().y, view);
+        observer->rotate(Eigen::Quaternionf::FromTwoVectors(oldPickRay.value(), newPickRay));
+    }
+
+    if ((renderer->getRenderFlags() & Renderer::ShowAutoMag))
```
Comment (375gnu) [inline src/celestia/celestiacore.cpp:2303]:
```suggestion
    if ((renderer->getRenderFlags() & Renderer::ShowAutoMag) != 0)
```

File: src/celestia/celestiacore.cpp [hunk 3]
```diff
@@ -2267,15 +2254,59 @@ Eigen::Vector3f CelestiaCore::getPickRay(float x, float y, const celestia::View
...
+
+    projectionMode->setSize(windowWidth, windowHeight);
+    return pickRay;
+}
+
+void CelestiaCore::updateFOV(float newFOV, std::optional<MouseLocation> focus, const celestia::View *view)
+{
+    float minFOV = renderer->getProjectionMode()->getMinimumFOV();
+    float maxFOV = renderer->getProjectionMode()->getMaximumFOV();
+    Observer *observer = view->getObserver();
```
Comment (375gnu) [inline src/celestia/celestiacore.cpp:2283]: move this definition closer to its usage

**Review by 375gnu — CHANGES_REQUESTED** [review]

File: src/celestia/celestiacore.cpp [hunk 4]
```diff
@@ -753,9 +751,7 @@ void CelestiaCore::pinchUpdate(float focusX, float focusY, float scale)
     const View *view = viewManager->activeView();
     float fov = view->getObserver()->getFOV();
 
-    MouseLocation focus;
-    focus.x = focusX;
-    focus.y = focusY;
+    Eigen::Vector2f focus { focusX, focusY };
```
Comment (375gnu) [inline src/celestia/celestiacore.cpp:754]: please use `{}` only in class declarations.

**Review by 375gnu — APPROVED** [review]

**Pull request merged**